core/DefaultRouteHandler.ts

```diff
@@ -0,0 +1,167 @@
+
+    constructor(public routeId: string, public routeUrl: string, public route: RouteRegistration, public decoratedMethod: Function, public middlewares: RequestHandler[]) {
+        this.wildcardCount = this.routeUrl.split('*').length - 1;
+        this.segmentCount = this.routeUrl.split('/').length;
+    }
+}
+
+@injectable()
+export class DefaultRouteHandler implements RouteHandler {
+    private routes: {[id: string]: RegistrationHelper} = {};
+
+    constructor(@inject(IoCSymbols.paramHandler) private paramHandler: ParamHandler) {
+    }
+
+    public addRoutes(controllerRegistration: ControllerRegistration, url: string): void {
+        let routes = Reflect.getOwnMetadata(ROUTES_KEY, controllerRegistration.controller) || [],
+            instance = new controllerRegistration.controller();
+
+        for (let route of routes) {
+            let ctrlTarget = controllerRegistration.controller,
+                routeTarget = ctrlTarget.prototype,
+                routeUrl = url + [controllerRegistration.prefix, route.path].filter(Boolean).join('/'),
+                returnType = Reflect.getMetadata('design:returntype', routeTarget, route.propertyKey),
+                middlewares = [...controllerRegistration.middlewares, ...route.middlewares],
+                method = route.descriptor.value;
+
+            let index;
+            if ((index = routeUrl.lastIndexOf('~')) > -1) {
+                routeUrl = routeUrl.substring(index + 1);
+            }
+
+            if (routeUrl.length > 1) {
+                routeUrl = '/' + routeUrl.split('/').filter(Boolean).join('/');
+            }
+
+            let routeId = routeUrl + route.method.toString();
+
+            if (this.routes[routeId]) {
+                throw new DuplicateRouteDeclarationError(routeUrl, route.method);
+            }
+
+            if (route.method === RouteMethod.Head && returnType !== Boolean) {
+                throw new HeadHasWrongReturnTypeError();
+            }
+
+            let params = this.paramHandler.getParamsForRoute(routeTarget, route.propertyKey),
+                hasResponseParam = params.some(p => p.paramType === ParamType.Response);
+
+            let decoratedMethod = (request: Request, response: Response, next) => {
+                let errorHandler: ControllerErrorHandler = Reflect.getMetadata(ERRORHANDLER_KEY, ctrlTarget),
+                    paramValues = [];
+
+                if (!errorHandler) {
+                    errorHandler = new ControllerErrorHandler();
+                }
+
+                let handleError = (error: any) => {
+                    if (!(error instanceof Error)) {
+                        error = new GenericRouteError(route.propertyKey, error);
+                    }
+                    errorHandler.handleError(instance, request, response, error);
+                };
+
+                try {
+                    paramValues = this.paramHandler.getParamValuesForRequest(params, request, response);
```

Comment:
I know there was a reason to not include this in the try-block below... but what was it again? :)